lib/mod/cache/fetch.go

```diff
@@ -47,10 +55,70 @@ func fetch(lang, mod, ver string) error {
 	switch remote {
 	case "github.com":
 		return fetchGitHub(lang, owner, repo, tag)
-
 	default:
-		return fmt.Errorf("Unknown remote: %q in %s", remote, mod)
+		return fetchGit(lang, remote, owner, repo, tag)
+	}
+}
+
+func fetchGit(lang, remote, owner, repo, tag string) error {
+	FS := memfs.New()
+
+	gco := &git.CloneOptions{
+		URL:   fmt.Sprintf("https://%s/%s/%s", remote, owner, repo),
+		Depth: 1,
+	}
+	if tag != "v0.0.0" {
+		gco.ReferenceName = plumbing.NewTagReferenceName(tag)
+		gco.SingleBranch = true
+	}
+
+	if _, err := git.Clone(memory.NewStorage(), FS, gco); err != nil {
+		if err != transport.ErrAuthenticationRequired {
+			return err
+		}
+
+		// Needs auth
+		fmt.Println("Private repo, trying to authenticate with ssh")
+		newRemote, auth, err := getSSHAuth(remote)
+		if err != nil {
+			return err
+		}
+		gco.URL = fmt.Sprintf("%s:%s/%s", newRemote, owner, repo)
+		gco.Auth = auth
+
+		if _, err := git.Clone(memory.NewStorage(), FS, gco); err != nil {
+			return err
+		}
 	}
+
+	if err := Write(lang, remote, owner, repo, tag, FS); err != nil {
+		return fmt.Errorf("While writing to cache\n%w\n", err)
+	}
+
+	return nil
+}
+
+func getSSHAuth(remote string) (string, *ssh.PublicKeys, error) {
+	pk, err := ssh_config.GetStrict(remote, "IdentityFile")
+	if err != nil {
+		return "", nil, err
+	}
+	if strings.HasPrefix(pk, "~") {
```

Comment:
Would this work for Windows? Didn't think that `~` worked there.

I should be able to test it.

Reply:
Didn't have the chance to try on Windows. I'm not really a Windows user so it's really possible that we'll have to tweak this a little bit in order to fit in a standard Windows way of retrieving the right ssh key.